src/main/resources/com/google/api/codegen/go/mock.snip

```diff
@@ -120,111 +125,181 @@
 
 @private clientStreamMethod(impl, method)
     func (s *{@impl.name}) {@method.name}(stream {@method.streamHandleTypeName}) error {
-        if s.err != nil {
-            return s.err
-        }
         for {
             if req, err := stream.Recv(); err == io.EOF {
-                return stream.SendAndClose(s.resps[0].({@method.responseTypeName}))
+                break
             } else if err != nil {
                 return err
             } else {
                 s.reqs = append(s.reqs, req)
             }
         }
+        if s.err != nil {
+            return s.err
+        }
+        return stream.SendAndClose(s.resps[0].({@method.responseTypeName}))
     }
 @end
 
 @private bidiMethod(impl, method)
     func (s *{@impl.name}) {@method.name}(stream {@method.streamHandleTypeName}) error {
+        for {
+            if req, err := stream.Recv(); err == io.EOF {
+                break
+            } else if err != nil {
+                return err
+            } else {
+                s.reqs = append(s.reqs, req)
+            }
+        }
         if s.err != nil {
             return s.err
         }
-
-        ch := make(chan error, 2)
-        go func() {
-            for {
-                if req, err := stream.Recv(); err == io.EOF {
-                    ch <- nil
-                    return
-                } else if err != nil {
-                    ch <- err
-                    return
-                } else {
-                    s.reqs = append(s.reqs, req)
-                }
-            }
-        }()
-        go func() {
-            for _, v := range s.resps {
-                if err := stream.Send(v.({@method.responseTypeName})); err != nil {
-                    ch <- err
-                    return
-                }
+        for _, v := range s.resps {
+            if err := stream.Send(v.({@method.responseTypeName})); err != nil {
+                return err
             }
-            ch <- nil
-        }()
-
-        // Doesn't really matter which one we get.
-        err := <-ch
-        if err2 := <-ch; err == nil {
-            err = err2
         }
-        return err
+        return nil
     }
 @end
 
-@private testError(test)
-    func {@test.nameWithException}(t *testing.T) {
-        errCode := codes.Internal
-        {@test.mockServiceVarName}.err = grpc.Errorf(errCode, "test error")
+@private testFunc(test)
```

Comment:
this implies a single test function, but there are two - maybe call this `testFunctions`